ambry-router/src/main/java/com/github/ambry/router/OperationController.java

```diff
@@ -0,0 +1,647 @@
+/*
+ * Copyright 2021 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.account.AccountService;
+import com.github.ambry.clustermap.ClusterMap;
+import com.github.ambry.clustermap.DataNodeId;
+import com.github.ambry.commons.Callback;
+import com.github.ambry.commons.ResponseHandler;
+import com.github.ambry.config.RouterConfig;
+import com.github.ambry.messageformat.BlobProperties;
+import com.github.ambry.network.NetworkClient;
+import com.github.ambry.network.NetworkClientFactory;
+import com.github.ambry.network.RequestInfo;
+import com.github.ambry.network.ResponseInfo;
+import com.github.ambry.notification.NotificationSystem;
+import com.github.ambry.protocol.GetOption;
+import com.github.ambry.protocol.RequestOrResponse;
+import com.github.ambry.protocol.RequestOrResponseType;
+import com.github.ambry.quota.QuotaChargeCallback;
+import com.github.ambry.utils.Time;
+import com.github.ambry.utils.Utils;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * OperationController is the scaling unit for the NonBlockingRouter. The NonBlockingRouter can have multiple
+ * OperationControllers. Any operation submitted to the NonBlockingRouter will be submitted to one of the
+ * OperationControllers. A worker thread (the RequestResponseHandler thread) will poll The OperationController for
+ * requests to be sent and will notify it on receiving responses. The OperationController in turn makes use of the
+ * {@link PutManager}, {@link GetManager} and {@link DeleteManager} to perform puts, gets and deletes,
+ * respectively. A {@link NetworkClient} is used to interact with the network.
+ */
+public class OperationController implements Runnable {
```

Comment:
You don't delete this class from NonBlockingRouter.java and also update the places where it is referenced?

Reply:
how do we do the quota checking and charging the operation  controller? or can we just do it in the NetworkClient?

Reply:
As discussed we do need quota checking in OperationController because there is a lot of business logic involved in quota checking that shouldn't be leaked to network modules.